lib/Analysis/GateAnalysis.cc

```diff
@@ -230,15 +223,32 @@ void GateAnalysisImpl::processPhi(PHINode *PN, Instruction *insertionPt) {
 
       // Or the successors unconditionally flows to an already processed block.
       // Note that there can be at most one such block.
-      for (auto &BlockValuePair : flowingValues) {
-        if (m_PDT.dominates(BlockValuePair.first, S)) {
-          SuccToVal[S] = BlockValuePair.second;
-          GSA_LOG(errs() << "2) SuccToVal[" << S->getName()
-                         << "] = postdom for cd "
-                         << BlockValuePair.first->getName() << ": "
-                         << SuccToVal[S]->getName() << "\n");
+      bool found = false;
+      BasicBlock *bb =  S;
+      auto *node = m_PDT.getNode(bb)->getIDom();
+      while (node != nullptr) {
+        auto it = flowingValues.find(bb);
+        if (it != flowingValues.end()) {
+          SuccToVal[S] = it->second;
+          found = true;
           break;
         }
+        bb = node->getBlock();
+        node = m_PDT.getNode(bb)->getIDom();
+      }
+      if (node == nullptr) {
```

Comment:
Can this loop be restructured so that we don't have to manually do one more iteration at the end? It seems you only really need to collect all the basic blocks along the postdom chain.

Reply:
Can you comment what this case means? `node == nullptr` should happen when you traverse the postdominance chain all the way up to the virtual root, right?

Reply:
Changed the loop condition and removed the condition. Shouldn't matter anymore.